### PR TITLE
New servus extractor logic #1076

### LINKED ISSUE #1076: [Broken] Servus TV
## Checklist
- [x] I'm reporting a broken site support
- [x] I've verified that I'm running yt-dlp version **2021.09.25**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log
```
./yt-dlp --verbose https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12
[debug] User config file: /home/user/.config/yt-dlp/config
[debug] User config: ['--restrict-filenames', '--no-mtime', '--compat-options', 'filename']
[debug] Command-line config: ['--verbose', 'https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] yt-dlp version 2021.09.25 (zip)
[debug] Compatibility options: filename
[debug] Python version 3.8.10 (CPython 64bit) - Linux-5.4.0-86-generic-x86_64-with-glibc2.29
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4, rtmpdump 2.4
[debug] Optional libraries: Crypto, keyring, sqlite
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12
[generic] aa-25fhsqdp51w12: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] aa-25fhsqdp51w12: Downloading webpage
[generic] aa-25fhsqdp51w12: Extracting information
ERROR: [generic] Unsupported URL: https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12
Traceback (most recent call last):
  File "./yt-dlp/yt_dlp/extractor/common.py", line 585, in extract
    ie_result = self._real_extract(url)
  File "./yt-dlp/yt_dlp/extractor/generic.py", line 3712, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12
Traceback (most recent call last):
  File "./yt-dlp/yt_dlp/extractor/common.py", line 585, in extract
    ie_result = self._real_extract(url)
  File "./yt-dlp/yt_dlp/extractor/generic.py", line 3712, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1227, in wrapper
    return func(self, *args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1252, in __extract_info
    ie_result = ie.extract(url)
  File "./yt-dlp/yt_dlp/extractor/common.py", line 601, in extract
    raise ExtractorError(
yt_dlp.utils.ExtractorError: [generic] Unsupported URL: https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12
```

```
./yt-dlp --verbose https://www.servustv.com/videos/aa-25fhsqdp51w12
[debug] User config file: /home/user/.config/yt-dlp/config
[debug] User config: ['--restrict-filenames', '--no-mtime', '--compat-options', 'filename']
[debug] Command-line config: ['--verbose', 'https://www.servustv.com/videos/aa-25fhsqdp51w12']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] yt-dlp version 2021.09.25 (zip)
[debug] Compatibility options: filename
[debug] Python version 3.8.10 (CPython 64bit) - Linux-5.4.0-86-generic-x86_64-with-glibc2.29
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4, rtmpdump 2.4
[debug] Optional libraries: Crypto, keyring, sqlite
[debug] Proxy map: {}
[debug] [Servus] Extracting URL: https://www.servustv.com/videos/aa-25fhsqdp51w12
[Servus] AA-25FHSQDP51W12: Downloading token
ERROR: [Servus] aa-25fhsqdp51w12: Unable to download JSON metadata: HTTP Error 401:  (caused by <HTTPError 401: ''>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.
  File "./yt-dlp/yt_dlp/extractor/common.py", line 694, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3256, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
```

## Description
Servus TV is not working since a couple of months. For a short time it was possible to modify the URL, e.g. from the original URL 
```https://www.servustv.com/unterhaltung/v/aa-25fhsqdp51w12``` to ```https://www.servustv.com/videos/aa-25fhsqdp51w12```
but this is not working any longer. Both URLs are shown in the verbose log above.

### PR DESCRIPTION
**Pull request by FrankZ85**
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

DESCRIPTION

Fixes #1076

### CODE REVIEW

File: yt_dlp/extractor/servus.py
Comment (Ashish0804) [inline yt_dlp/extractor/servus.py:57]:
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            video['videoUrl'], video_id, 'mp4', entry_protocol='m3u8_native',
            m3u8_id='hls')
```

File: yt_dlp/extractor/servus.py
Comment (gamer191) [inline yt_dlp/extractor/servus.py:48]: The blank line at line 48 should be removed, in order to make the automated linter happy
```suggestion

```

**Comment by StefanLobbenmeier** [comment]
I just ran a successful test of this on my machine. The old version fails like this:
```
➜  Downloads ./yt-dlp_macos_legacy https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
[generic] aa7u8970e8drj41nx6z0: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] aa7u8970e8drj41nx6z0: Downloading webpage
[generic] aa7u8970e8drj41nx6z0: Extracting information
ERROR: Unsupported URL: https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
```

After applying this patch I could successfully download:
➜  Downloads ./yt-dlp_macos_legacy-2 https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
```
[Servus] AA7U8970E8DRJ41NX6Z0: Downloading video JSON
[Servus] AA7U8970E8DRJ41NX6Z0: Downloading m3u8 information
[info] AA7U8970E8DRJ41NX6Z0: Downloading 1 format(s): hls-6101+hls-audio-stereo-aac-English
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 69
[download] Destination: Die Weltmeister-Werkstatt 1 [AA7U8970E8DRJ41NX6Z0].fhls-6101.mp4
[download]   4.6% of ~122.63MiB at    3.04MiB/s ETA 00:31 (frag 3/69)^C
ERROR: Interrupted by user
```

**Comment by StefanLobbenmeier** [comment]
I might have found a non working example: https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/, but it could also be my build

```
~/Downloads ❯❯❯ ./yt-dlp_macos_legacy-2 -v https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/
[debug] Command-line config: ['-v', 'https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.09.140503 [9e642a8] (darwin_exe)
[debug] Python 3.10.5 (CPython 64bit) - macOS-12.3-x86_64-i386-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [Servus] Extracting URL: https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/
[Servus] AAVUYK29A8FSW6CSX4V7: Downloading video JSON
[Servus] AAVUYK29A8FSW6CSX4V7: Downloading m3u8 information
ERROR: 'bytes' object has no attribute 'encode'
Traceback (most recent call last):
  File "yt_dlp/YoutubeDL.py", line 1425, in wrapper
  File "yt_dlp/YoutubeDL.py", line 1495, in __extract_info
  File "yt_dlp/extractor/common.py", line 640, in extract
  File "yt_dlp/extractor/servus.py", line 55, in _real_extract
  File "yt_dlp/extractor/common.py", line 2132, in _extract_m3u8_formats_and_subtitles
  File "yt_dlp/extractor/common.py", line 834, in _download_webpage_handle
  File "yt_dlp/extractor/common.py", line 773, in _request_webpage
  File "yt_dlp/extractor/common.py", line 739, in _create_request
  File "yt_dlp/utils.py", line 752, in sanitized_Request
  File "yt_dlp/utils.py", line 2989, in escape_url
AttributeError: 'bytes' object has no attribute 'encode'
```

**Comment by StefanLobbenmeier** [comment]
The example https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/ actually doesn’t play the full movie in the browser either, just a trailer saying that it will run at 20:15

So it might actually work if we try again later, but it would be nicer to have a dedicated error message or download the trailer? Not sure, but not a huge issue if it ends up working later I think

**Comment by pukkandan** [comment]
How is it that there are so many people willing to make PR for this site, but no one want to address reviews... https://github.com/yt-dlp/yt-dlp/pull/1045 https://github.com/yt-dlp/yt-dlp/pull/2748 https://github.com/yt-dlp/yt-dlp/pull/1498

**Comment by FrankZ85** [comment]
> How is it that there are so many people willing to make PR for this site, but no one want to address reviews... #1045 #2748 #1498

I have just added some more functionality according to my research on the site, and addressed the reviews.

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:59]:
```suggestion
            video['videoUrl'], video_id, 'mp4', m3u8_id='hls')
```
It is the default

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:97]: raise ExtractorError. Same below. When the error does not indicate a bug (eg: `only playable from ...`), set `expected=true`

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:87]: I don't think a description is useful enough to make an additional API call for it. But I leave it to your discretion

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:93]:
```suggestion
        return traverse_obj(
            info, 'stv_long_description', 'stv_short_description') or video.get('description')
```

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:76]: typo?

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:77]:
```suggestion
            'season': season,
```

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:52]: fstrings are prefered over `%`, but I won't force you

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:44]: The existing tests are meant to test those URL schemes. Don't replace them

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:86]: Use single quotes everywhere

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:93]: Actually, passing `video` here only to do a `.get` on it seems unnecessary. Do this instead:
```suggestion
    def _get_description(self, video_id):
        info = self._download_json(
            f'https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/{video_id}?fieldset=page',
            video_id, fatal=False)
        return traverse_obj(info, 'stv_long_description', 'stv_short_description')
```
and above
```py
            'description': self._get_description(video_id) or video.get('description'),
```

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:104]: format it like:
```py
                    raise ExtractorError(
                        f'Only playable from {details["minEveningHour"]}:00 to {details["maxMorningHour"]}:00',
                        expected=True)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by gamer191** [comment]
@flashdagger after this gets merged are you planning on maintaining your servustv plugin?

File: yt_dlp/extractor/servus.py
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:52]: Use `query` param

**Comment by pukkandan** [comment]
@gamer191 It implements much more functionality afair

**Comment by StefanLobbenmeier** [comment]
Is there anything I can do to make this PR merged? @FrankZ85 would you mind if I took it over

**Comment by pukkandan** [comment]
> Is there anything I can do to make this PR merged?

If OP is not going to address the review, you can make a new PR addressing the current comments.

When making PR, make sure you are willing and able to address future reviews. I already reviewed 4 different PRs for the same issue, all the contributors abandoning the PR midway! If you don't want to deal with our conventions/requirements, distribute it as a plugin instead and pls don't waste time of all involved

**Comment by StefanLobbenmeier** [comment]
> If OP is not going to address the review, you can make a new PR addressing the current comments.

We waited another 2 weeks after pinging him, so I am assuming he will not continue.

I will start by making a copy of this PR and replicate your PR comments, and then fix them in my version to minimize the work on your side 😄 as for dealing with new review comments I will deal with them

**Pull request closed without merging**
